### DISCUSSION TITLE
How does the tailwindcss postcss plugin watch for new changes to documents?

### DISCUSSION THREAD

**Discussion by redonkulus** [tailwindlabs/tailwindcss#11499] [Help] [answered]
I'm using Next.js and the Tailwind postcss plugin to build the CSS. I'm curious how the tailwind post-css plugins knows there are changes to the document and re-runs tailwind. I'm trying to build my own post-css plugin but can't seem to figure out how to watch for changes to the document? A pointer to the code in tailwind would help me figure this out.

**Comment by wongjn** [comment]
I feel like this might be a key line:

https://github.com/tailwindlabs/tailwindcss/blob/e4af07f7ad3e61a977f74e31f0cf856f638c0bf2/src/plugin.js#L31

Perhaps along with this:

https://github.com/tailwindlabs/tailwindcss/blob/e4af07f7ad3e61a977f74e31f0cf856f638c0bf2/src/processTailwindFeatures.js#L24-L37

Also, as per [the PostCSS documentation on Writing a Plugin](https://postcss.org/docs/writing-a-postcss-plugin#step-change-nodes):

> If your plugin depends on another file, you can attach a message to result to signify to runners (webpack, Gulp etc.) that they should rebuild the CSS when this file changes:
>
> ```js
> AtRule: {
>   import: (atRule, { result }) {
>     const importedFile = parseImport(atRule)
>     result.messages.push({
>       type: 'dependency',
>       plugin: 'postcss-import',
>       file: importedFile,
>       parent: result.opts.from
>     })
>   }
> }
> ```
>
> If the dependency is a directory you should use the dir-dependency message type instead:
>
> ```js
> result.messages.push({
>   type: 'dir-dependency',
>   plugin: 'postcss-import',
>   dir: importedDir,
>   parent: result.opts.from
> })
> ```

**Reply by redonkulus** [reply]
Thanks, I saw that too but had no idea what it does or what its used for.

**Reply by wongjn** [reply]
I believe it may be used to know there are changes to the document and re-run Tailwind.

**Reply by redonkulus** [reply]
What triggers that code to run? From what I see in post-css plugins, there is the Document() or Root() event that runs when post-css is triggered. I'm not sure which event is used for subsequent document/root updates. Its also possible its Next.js not triggering this event too. But Tailwind is able to run based on updates to the next.js app files, so I'm not sure what's happening here.

**Reply by wongjn** [reply, marked as answer]
[As stated above](https://github.com/tailwindlabs/tailwindcss/discussions/11499#discussioncomment-6297366):

> Also, as per [the PostCSS documentation on Writing a Plugin](https://postcss.org/docs/writing-a-postcss-plugin#step-change-nodes):
> 
> > If your plugin depends on another file, you can attach a message to result to signify to runners (webpack, Gulp > etc.) that they should rebuild the CSS when this file changes:
> >
> > ```js
> > AtRule: {
> >   import: (atRule, { result }) {
> >     const importedFile = parseImport(atRule)
> >     result.messages.push({
> >       type: 'dependency',
> >       plugin: 'postcss-import',
> >       file: importedFile,
> >       parent: result.opts.from
> >     })
> >   }
> > }
> > ```
> >
> > If the dependency is a directory you should use the dir-dependency message type instead:
> >
> > ```js
> > result.messages.push({
> >   type: 'dir-dependency',
> >   plugin: 'postcss-import',
> >   dir: importedDir,
> >   parent: result.opts.from
> > })
> > ```

**Reply by redonkulus** [reply]
Using `result.messages.push` made it work, thanks for finding that for me!
